language/move-vm/runtime/src/native_extensions.rs

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use better_any::{Tid, TidAble, TidExt};
+use std::{any::TypeId, collections::HashMap};
+
+/// A data type to represent a heterogeneous collection of extensions which are available to
+/// native functions. A value to this is passed into the session function execution.
+///
+/// The implementation uses the crate `better_any` which implements a version of the `Any`
+/// type, called `Tid<`a>`, which allows for up to one lifetime parameter. This
+/// avoids that extensions need to have `'static` lifetime, which `Any` requires. In order to make a
+/// struct suitable to be a 'Tid', use `#[derive(Tid)]` in the struct declaration. (See also
+/// tests at the end of this module.)
+#[derive(Default)]
+pub struct NativeContextExtensions<'a> {
+    map: HashMap<TypeId, Box<dyn Tid<'a>>>,
+}
+
+impl<'a> NativeContextExtensions<'a> {
+    pub fn add<T: TidAble<'a>>(&mut self, ext: T) {
+        assert!(
```

Comment:
should this maybe return a `VMResult`? That way we can handle the assert gracefully?

I know we don't do this everywhere in the VM, but we try to do it where we can (I hope)

Reply:
And it shouldn't be that much more painful for anyone using this, they will likely just have to slap a `?` on the call

Reply:
That's actually an interesting thing we should discuss. At Google I learned that fail fast and assertive in code which correctness does not depend on user input is preferred. So I would say that

- It should be VMResult whenever the behavior depends e.g. on bytecode input.
- It should be actually assert if it does not depend on bytecode. (This is also easier to debug BTW.)

We can continue to discuss this. Landing for now.

/land